File: app/core/logging.py
import logging
import re
from pythonjsonlogger import jsonlogger

SENSITIVE_PATTERNS = [
    re.compile(r"(?i)(password|passwd|pwd)\s*[:=]\s*\S+"),
    re.compile(r"(?i)(api[_-]?key|token|secret)\s*[:=]\s*\S+"),
    re.compile(r"(?i)authorization\s*[:=]\s*bearer\s+\S+"),
]


def sanitize_text(value: str) -> str:
    sanitized = value
    for pattern in SENSITIVE_PATTERNS:
        sanitized = pattern.sub("[REDACTED]", sanitized)
    return sanitized


class SanitizeFilter(logging.Filter):
    def filter(self, record: logging.LogRecord) -> bool:
        if isinstance(record.msg, str):
            record.msg = sanitize_text(record.msg)
        return True


def configure_logging(level: str = "INFO") -> None:
    root = logging.getLogger()
    root.handlers.clear()
    root.setLevel(level.upper())

    handler = logging.StreamHandler()
    formatter = jsonlogger.JsonFormatter("%(asctime)s %(levelname)s %(name)s %(message)s")
    handler.setFormatter(formatter)
    handler.addFilter(SanitizeFilter())
    root.addHandler(handler)

File: src/core/logging.js
const pino = require("pino");

const SENSITIVE_PATTERNS = [
  /(password|passwd|pwd)\s*[:=]\s*\S+/gi,
  /(api[_-]?key|token|secret)\s*[:=]\s*\S+/gi,
  /authorization\s*[:=]\s*bearer\s+\S+/gi,
];

function sanitizeText(value) {
  if (typeof value !== "string") return value;
  let sanitized = value;
  for (const pattern of SENSITIVE_PATTERNS) {
    sanitized = sanitized.replace(pattern, "[REDACTED]");
  }
  return sanitized;
}

function sanitizeObject(obj) {
  if (!obj || typeof obj !== "object") return obj;
  const clone = Array.isArray(obj) ? [...obj] : { ...obj };
  for (const key of Object.keys(clone)) {
    const value = clone[key];
    if (typeof value === "string") clone[key] = sanitizeText(value);
  }
  return clone;
}

function createLogger(level = "info") {
  return pino({ level });
}

module.exports = { createLogger, sanitizeText, sanitizeObject };
